File: scripts/convert-pdf-to-images.js
import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { execSync } from 'child_process';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const PDF_PATH = path.join(__dirname, '../public/products/product-catalogue-2026.pdf');
const OUTPUT_DIR = path.join(__dirname, '../public/products/catalogue-pages');

async function convertPdfToImages() {
  try {
    // Create output directory
    if (!fs.existsSync(OUTPUT_DIR)) {
      fs.mkdirSync(OUTPUT_DIR, { recursive: true });
    }

    console.log(`📖 Loading PDF from: ${PDF_PATH}`);
    console.log(`🖼️  Converting PDF to images using ImageMagick...`);

    // Use ImageMagick convert command to convert PDF to PNG
    // This requires ImageMagick to be installed on the system
    try {
      const command = `magick "${PDF_PATH}" -density 150 -quality 90 "${OUTPUT_DIR}/page-%%02d.png"`;
      console.log(`Running: ${command}`);
      execSync(command, { stdio: 'inherit' });
      
      console.log(`\n✅ Successfully converted PDF to images!`);
      console.log(`📁 Images saved to: ${OUTPUT_DIR}`);
    } catch (error) {
      // If ImageMagick is not available, try alternative methods
      console.log('\n⚠️  ImageMagick not found. Trying alternative method...');
      
      // Try with pdftoppm (from poppler-utils) if available
      try {
        const command = `pdftoppm -png -r 150 "${PDF_PATH}" "${OUTPUT_DIR}/page"`;
        console.log(`Running: ${command}`);
        execSync(command, { stdio: 'inherit' });
        
        // Rename files to match expected format (page-01.png, page-02.png, etc)
        const files = fs.readdirSync(OUTPUT_DIR).filter(f => f.startsWith('page-') && f.endsWith('.png'));
        files.forEach((file, index) => {
          const oldPath = path.join(OUTPUT_DIR, file);
          const pageNum = String(index + 1).padStart(2, '0');
          const newPath = path.join(OUTPUT_DIR, `page-${pageNum}.png`);
          if (oldPath !== newPath && fs.existsSync(oldPath)) {
            fs.renameSync(oldPath, newPath);
          }
        });
        
        console.log(`\n✅ Successfully converted ${files.length} pages!`);
        console.log(`📁 Images saved to: ${OUTPUT_DIR}`);
      } catch (pdferror) {
        console.error('\n❌ Neither ImageMagick nor pdftoppm found!');
        console.error('\nTo fix this, install one of the following:');
        console.error('  • ImageMagick: https://imagemagick.org/script/download.php');
        console.error('  • Poppler (pdftoppm): https://github.com/oschwartz10612/poppler-windows/releases/');
        process.exit(1);
      }
    }
  } catch (error) {
    console.error('❌ Error converting PDF to images:', error.message);
    process.exit(1);
  }
}

convertPdfToImages();

File: scripts/convert-pdf-to-images.py
#!/usr/bin/env python3
"""Convert PDF pages to PNG images for the flipbook display."""

import sys
from pathlib import Path

import pymupdf


SCRIPT_DIR = Path(__file__).parent
PROJECT_ROOT = SCRIPT_DIR.parent
PDF_PATH = PROJECT_ROOT / "public" / "products" / "product-catalogue-2026.pdf"
OUTPUT_DIR = PROJECT_ROOT / "public" / "products" / "catalogue-pages"
PAGE_COUNT = 26
# High-resolution source images keep small catalogue text sharp when zoomed
# or displayed in the full-screen reader.
ZOOM = 2.5


def convert_pdf_to_images():
    """Render the catalogue's first PAGE_COUNT pages as PNG files."""
    try:
        OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

        if not PDF_PATH.exists():
            print(f"PDF file not found: {PDF_PATH}")
            sys.exit(1)

        print(f"Loading PDF from: {PDF_PATH}")
        pdf_document = pymupdf.open(str(PDF_PATH))
        num_pages = min(pdf_document.page_count, PAGE_COUNT)
        print(f"Converting {num_pages} pages to images...")

        for page_num in range(num_pages):
            page = pdf_document[page_num]
            pixmap = page.get_pixmap(matrix=pymupdf.Matrix(ZOOM, ZOOM), alpha=False)
            filename = f"page-{page_num + 1:02d}.png"
            pixmap.save(str(OUTPUT_DIR / filename))
            print(f"  Converted page {page_num + 1}/{num_pages}: {filename}")

        pdf_document.close()
        print(f"Successfully converted {num_pages} pages.")
        print(f"Images saved to: {OUTPUT_DIR}")
    except ImportError as error:
        print(f"Missing dependency: {error}")
        print("Install it with: python -m pip install PyMuPDF")
        sys.exit(1)
    except Exception as error:
        print(f"Error converting PDF to images: {error}")
        sys.exit(1)


if __name__ == "__main__":
    convert_pdf_to_images()
